### PR TITLE
Feature: A new contrastive loss (Barlow Twins)

### PR DESCRIPTION
**Pull request by Nimrais**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [ ] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/drive/1JCGTVvWlrIsLXMPRRRSWiAstSLic4nbA) with `latest` and `minimal` requirements?

## Description

Implementation of a negative-free contrastive loss from [Barlow Twins: Self-Supervised Learning via Redundancy Reduction](https://arxiv.org/abs/2103.03230).

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

**Comment by Nimrais** [comment]
I have run `pytest .` but [test_reid_pipeline](https://github.com/catalyst-team/catalyst/blob/master/tests/catalyst/callbacks/test_metric.py) fails. I am sure it's not dependent on my code but it can be connected with my machine (apple M1). The test fails on `assert runner.loader_metrics["cmc01"] > 0.7`.

File: catalyst/contrib/nn/criterion/contrastive.py [hunk 1]
```diff
@@ -135,8 +135,77 @@ def forward(self, embeddings_pred, embeddings_true) -> torch.Tensor:
...
+        >>> import torch
+        >>> from torch.nn import functional as F
+        >>> from catalyst.contrib.nn import BarlowTwinsLoss
+        >>>
+        >>> embeddings_left = torch.tensor([[1., 0.0],[0.0, 1.0]], requires_grad=True)
+        >>> embeddings_right = torch.tensor([[1., 0.0],[0.0, 1.0]], requires_grad=True)
+        >>> criterion = BarlowTwinsLoss(lmbda = 1)
+        >>> criterion_val = criterion(embeddings_left, embeddings_right).item()
+        >>> abs(1 - criterion_val) < 1e-6
+        True
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/contrastive.py]: could you please use `.. code-block:: python`? 
like https://github.com/catalyst-team/catalyst/blob/master/catalyst/metrics/functional/_f1_score.py#L33#L48
this format is much better for online documentation

File: catalyst/contrib/nn/criterion/contrastive.py [hunk 2]
```diff
@@ -135,8 +135,77 @@ def forward(self, embeddings_pred, embeddings_true) -> torch.Tensor:
         return loss
 
 
+class BarlowTwinsLoss(nn.Module):
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/contrastive.py:138]: could you please also drop a couple of tests for this module?
like https://github.com/catalyst-team/catalyst/blob/master/tests/catalyst/metrics/functional/test_accuracy.py#L48#L83

Comment (Scitator) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: just to check the correctness of the implementation ;)

**Comment by Scitator** [comment]
> I have run pytest . but test_reid_pipeline fails. I am sure it's not dependent on my code but it can be connected with my machine (apple M1). The test fails on assert runner.loader_metrics["cmc01"] > 0.7.

that's why we have stable GitHub actions tests :)

File: catalyst/contrib/nn/criterion/contrastive.py (hunk 1 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py]: Yes, sure.

File: catalyst/contrib/nn/criterion/contrastive.py (hunk 2 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: 👍  I will do it.

Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: [Done](ContrastivePairwiseEmbeddingLoss)

File: catalyst/contrib/nn/criterion/contrastive.py (hunk 1 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py]: [Done](https://github.com/catalyst-team/catalyst/pull/1259/commits/e064ca3698148a7264150b36412b8a48ba774dd5)

File: catalyst/contrib/nn/criterion/contrastive.py (hunk 2 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: Where should I put the tests?
I think here `tests/catalyst/contrib/nn/test_criterion.py` what do you think?

Comment (Scitator) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: LGTM 👍

Comment (Nimrais) [reply catalyst/contrib/nn/criterion/contrastive.py:138]: I think these [tests](https://github.com/Nimrais/catalyst/blob/barlow-twins/tests/catalyst/contrib/nn/test_criterion.py#L32-L108) are good enough.

File: catalyst/contrib/nn/criterion/contrastive.py [hunk 3]
```diff
@@ -135,8 +135,80 @@ def forward(self, embeddings_pred, embeddings_true) -> torch.Tensor:
...
+        criterion = BarlowTwinsLoss(lmbda = 1)
+        criterion_val = criterion(embeddings_left, embeddings_right).item()
+        abs(1 - criterion_val) < 1e-6
+        # True
+
+    .. _`Barlow Twins: Self-Supervised Learning via Redundancy Reduction`:
+        https://arxiv.org/abs/2103.03230
+    """
+
+    def __init__(self, lmbda=1.0, eps=1e-12):
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/contrastive.py]:
```suggestion
    def __init__(self, lambda=1.0, eps=1e-12):
```
?
